apps/grader-host/src/lib/receiveFromQueue.js

```diff
@@ -9,12 +9,63 @@ const {
   ChangeMessageVisibilityCommand,
   DeleteMessageCommand,
 } = require('@aws-sdk/client-sqs');
+const Sentry = require('@prairielearn/sentry');
+const { setTimeout: sleep } = require('node:timers/promises');
 
 const globalLogger = require('./logger');
 const { config } = require('./config');
 
 let messageSchema = null;
 
+/**
+ * @param {import('@aws-sdk/client-sqs').SQSClient} sqs
+ * @param {string} queueUrl
+ * @param {string} receiptHandle
+ * @param {number} timeout
+ * @returns {Promise<void>}
+ */
+async function changeVisibilityTimeout(sqs, queueUrl, receiptHandle, timeout) {
+  await sqs.send(
+    new ChangeMessageVisibilityCommand({
+      QueueUrl: queueUrl,
+      ReceiptHandle: receiptHandle,
+      VisibilityTimeout: timeout,
+    }),
+  );
+}
+
+/**
+ * @param {import('@aws-sdk/client-sqs').SQSClient} sqs
+ * @param {string} queueUrl
+ * @param {string} receiptHandle
+ */
+async function startHeartbeat(sqs, queueUrl, receiptHandle) {
+  const abortController = new AbortController();
+
+  // Run the first extension immediately before we start processing the job.
+  await changeVisibilityTimeout(sqs, queueUrl, receiptHandle, config.visibilityTimeout);
+
+  // We want this process to run in the background, so we don't await it.
+  // `extendVisibilityTimeout` will handle errors.
+  // eslint-disable-next-line no-floating-promise/no-floating-promise
+  (async () => {
+    while (!abortController.signal.aborted) {
+      await sleep(config.visibilityTimeoutHeartbeatIntervalSec * 1000, null, { ref: false });
+
+      if (abortController.signal.aborted) return;
+
+      try {
+        await changeVisibilityTimeout(sqs, queueUrl, receiptHandle, config.visibilityTimeout);
+      } catch (err) {
+        globalLogger.error('Error extending visibility timeout', err);
+        Sentry.captureException(err);
+      }
+    }
+  })();
```

Comment:
I resisted the urge to run off and build some more clever abstraction here, but man, I miss [Go's channels + goroutines](https://gobyexample.com/tickers)!